### PR TITLE
Add crypto module, responsible for generating and handling keys and the like

### LINKED ISSUE #37: Private child key derivation

### LINKED ISSUE #72: Public child key derivation

### LINKED ISSUE #82: build(npm): use @types/secp256k1 npm package
Remove `typings/secp256k1` and use instead [@types/secp256k1](https://www.npmjs.com/package/@types/secp256k1) npm package

### LINKED ISSUE #92: build(yarn): bignum failing to build
In node versions 7, 9 and 10 `bignum` is failing to build.

    The module '~/sheikah/node_modules/bignum/binding/bignum.node'

    was compiled against a different Node.js version using
    NODE_MODULE_VERSION 59. This version of Node.js requires
    NODE_MODULE_VERSION 64. Please try re-compiling or re-installing

### PR DESCRIPTION
**Pull request by kronolynx**
- [x] I have read and understood [CODE_OF_CONDUCT][code] document.
- [x] I have read and understood [docs/styleguide][style] document.
- [x] I have included tests for the changes in my PR. If not, I have included a
  rationale for why I haven't.
- [x] My code is formatted and is accepted by `yarn fmt-verify`.
- [x] **I understand my PR may be closed if it becomes obvious I didn't
  actually perform all of these steps.**

As demanded by #19

[code]: https://github.com/witnet/sheikah/blob/master/CODE_OF_CONDUCT.md
[style]: https://github.com/witnet/sheikah/blob/master/docs/styleguide.md

This closes #37, closes #72, closes #82, closes #92

### CODE REVIEW

**Comment by aesedepece** [comment]
We'll try to merge this later today. @kronolynx is doing some last minute fixes and getting everything ready for the big moment :muscle:

File: app/lib/crypto/seed.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+export const deriveSeedFromMnemonics = (mnemonics: string): Seed => {
+  if (Mnemonic.validate(mnemonics)) {
+    throw new Error("Invalid mnemonic")
+  }
+  const entropy = bip39.mnemonicToSeed(mnemonics)
+
+  return deriveSeedFromEntropy(entropy)
+}
+
+export const deriveSeedFromEntropy = (entropy: Buffer, hmacKey = "Witnet seed"): Seed => {
```
Comment (aesedepece) [inline app/lib/crypto/seed.ts]:
```ts
function fromEntropy(entropy: Buffer, hmacKey = "Witnet seed"): Seed

File: config/tslint.json [hunk 2]
```diff
@@ -19,7 +19,13 @@
     "jsdoc-format": true,
     "label-position": true,
     "linebreak-style": true,
-    "max-line-length": [true, 100],
+    "max-line-length": [
+      true,
+      {
+        "limit": 100,
+        "ignore-pattern": "^import |^export {(.*?)}| https?://"
```
Comment (aesedepece) [inline config/tslint.json]: Cool! :+1:

File: app/lib/crypto/keyPath.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  const toNumber = (value: string): number => {
+    const number = _.endsWith(value, "'") ?
+      hardened(_.parseInt(value.slice(0, -1))) : _.parseInt(value)
+    if (isNaN(number)) {
+      throw Error("Invalid number")
+    }
+
+    return number
+  }
+  const path1 = _.trimStart(_.trimStart(path, "m"), "/")
```
[the remark this answers is not in the document]
Comment (kronolynx) [reply app/lib/crypto/keyPath.ts]: they are not equivalent, I'm removing first the 'm'  then removing the first '\' if they are present

File: app/lib/crypto/constants.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import * as BigNum from "bignum"
+/**
```
Comment (aesedepece) [inline app/lib/crypto/constants.ts]: Put empty line here please :stuck_out_tongue:

File: app/lib/crypto/keyPath.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (aesedepece) [reply app/lib/crypto/keyPath.ts]: Can you provide a valid key path string for which these are not equivalent?

File: app/lib/crypto/constants.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+ * @type {BigNum}
+ */
+export const SECP256K1_N = BigNum.fromBuffer(
+  Buffer.from(
+    "FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141",
+    "hex"
+  )
+)
+
+export const HARDENED_KEY_INDEX = 0x80000000
```
Comment (aesedepece) [inline app/lib/crypto/constants.ts]: Love the usage of uppercase for invariants :heart_decoration:

File: app/lib/crypto/keyPath.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (kronolynx) [reply app/lib/crypto/keyPath.ts]: These tests fail:

m/0'/1
m/0'/1/2'
m/0'/1/2'/2
m/0'/1/2'/2/1000000000
m/0
m/0/2147483647'
m0_2147483647'/1
m0/2147483647'/1/2147483646'
m0/2147483647'/1/2147483646'/2

**Pull request merged**